File: Plugins/Orimath.Basics.View/ViewModels/InstructionArrowViewModel.cs
using Orimath.Controls;
using Orimath.Plugins;
using Orimath.FoldingInstruction;

namespace Orimath.Basics.View.ViewModels
{
    public class InstructionArrowViewModel : NotifyPropertyChanged
    {
        private readonly IViewPointConverter _pointConverter;

        private double _x1;
        public double X1 { get => _x1; set => SetValue(ref _x1, value); }

        private double _x2;
        public double X2 { get => _x2; set => SetValue(ref _x2, value); }

        private double _y1;
        public double Y1 { get => _y1; set => SetValue(ref _y1, value); }

        private double _y2;
        public double Y2 { get => _y2; set => SetValue(ref _y2, value); }

        private ArrowType _startType;
        public ArrowType StartType { get => _startType; set => SetValue(ref _startType, value); }

        private ArrowType _endType;
        public ArrowType EndType { get => _endType; set => SetValue(ref _endType, value); }

        private InstructionColor _color;
        public InstructionColor Color { get => _color; set => SetValue(ref _color, value); }

        private ArrowDirection _direction;
        public ArrowDirection Direction { get => _direction; set => SetValue(ref _direction, value); }

        public InstructionArrowViewModel(IViewPointConverter pointConverter)
        {
            _pointConverter = pointConverter;
        }

        public InstructionArrowViewModel(IViewPointConverter pointConverter, InstructionArrow model)
            : this(pointConverter)
        {
            SetModel(model);
        }

        public void SetModel(InstructionArrow model)
        {
            var point1 = _pointConverter.ModelToView(model.Line.Point1);
            var point2 = _pointConverter.ModelToView(model.Line.Point2);
            X1 = point1.X;
            Y1 = point1.Y;
            X2 = point2.X;
            Y2 = point2.Y;
            StartType = model.StartType;
            EndType = model.EndType;
            Color = model.Color;
            Direction = model.Direction;
        }
    }
}

File: Plugins/Orimath.Basics.View/ViewModels/InstructionArrowViewModel.fs
namespace Orimath.Basics.View.ViewModels
open Orimath.Controls
open Orimath.Plugins
open Orimath.FoldingInstruction
open ApplicativeProperty

type InstructionArrowViewModel(pointConverter: IViewPointConverter) =
    inherit NotifyPropertyChanged()
    member val X1 = Prop.value 0.0
    member val X2 = Prop.value 0.0
    member val Y1 = Prop.value 0.0
    member val Y2 = Prop.value 0.0
    member val StartType = Prop.value ArrowType.None
    member val EndType = Prop.value ArrowType.None
    member val Color = Prop.value InstructionColor.Black
    member val Direction = Prop.value ArrowDirection.Auto

    member this.SetModel(model: InstructionArrow) =
        let p1 = pointConverter.ModelToView(model.Line.Point1)
        let p2 = pointConverter.ModelToView(model.Line.Point2)
        this.X1.Value <- p1.X
        this.Y1.Value <- p1.Y
        this.X2.Value <- p2.X
        this.Y2.Value <- p2.Y
        this.StartType.Value <- model.StartType
        this.EndType.Value <- model.EndType
        this.Color.Value <- model.Color
        this.Direction.Value <- model.Direction

    static member Create(pointConverter, model) =
        let vm = InstructionArrowViewModel(pointConverter)
        vm.SetModel(model)
        vm
